draftjs_exporter/engines/string.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import absolute_import, unicode_literals
+
+from draftjs_exporter.engines.base import DOMEngine
+
+try:
+    # Python 3.2 and above.
+    from html import escape
+except ImportError:
+    import cgi
+
+    def escape(s):
+        # Force quote escaping in Python 2.7.
+        return cgi.escape(s, quote=True).replace('\'', '&#x27;')
+
+# http://w3c.github.io/html/single-page.html#void-elements
+# https://github.com/html5lib/html5lib-python/blob/0cae52b2073e3f2220db93a7650901f2200f2a13/html5lib/constants.py#L560
+VOID_ELEMENTS = [
+    'area',
+    'base',
+    'br',
+    'col',
+    'embed',
+    'hr',
+    'img',
+    'input',
+    'link',
+    'meta',
+    'param',
+    'source',
+    'track',
+    'wbr',
+]
+
+
+class DOM_STRING(DOMEngine):
+    """
+    string concat implementation of the DOM API.
+    """
+
+    @staticmethod
+    def create_tag(type_, attr=None):
+        return {
+            'type': type_,
+            'attr': attr,
+            'children': [],
+        }
+
+    @staticmethod
+    def append_child(elt, child):
+        # This check is necessary because the current wrapper_state implementation
+        # has an issue where it inserts elements multiple times.
+        # This must be skipped for text, which can be duplicated.
+        is_existing_ref = isinstance(child, dict) and child in elt['children']
+        if not is_existing_ref:
+            elt['children'].append(child)
+
+    @staticmethod
+    def render_attrs(attr):
+        attrs = [' {0}="{1}"'.format(a, escape(attr[a])) for a in attr]
```

Comment:
Always use the `%` operator for string formatting, it is 10 times faster than `str.format()`.
I know Python’s team has pushed a lot `str.format()` in the past few years, but this has changed and `%` is still very relevant, mostly because of its performance as well as similarity to other languages.

Reply:
For clarity, avoid adding a space before the attribute name. Instead, join using a space, then in the tag pattern in `DOM_STRING.render()`, add a space between tag name and attrs.

Reply:
That's very interesting, I didn't know that `%` was faster & still recommended.

For the spaces between attributes, it's a bit more complicated. The current implementation outputs the correct format when there are no attributes (eg. `<br/>` without any space within the tag), whereas what you are suggesting means there would need to be a check for whether to add the space between tag name and attribute list if there are attributes. While it's more semantically correct, it's also an extra operation.